### PR TITLE
Uefi decompress

### PR DESCRIPTION
**Pull request by BrentHoltsclaw**
Modified chipsec UEFI Decompress functionality to match edk2 Basetools VolInfo.

### CODE REVIEW

File: chipsec/hal/spi_uefi.py [hunk 1]
```diff
@@ -284,18 +284,48 @@ def build_efi_modules_tree( _uefi, fwtype, data, Size, offset, polarity ):
...
                             sec.children = build_efi_modules_tree( _uefi, fwtype, d, len(d), 0, polarity )
-                    #elif sec.Guid == FIRMWARE_VOLUME_GUID:
                     else:
                         sec.children = build_efi_model( _uefi, sec.Image[sec.HeaderSize:], fwtype )
-                elif sec.Type in (EFI_SECTION_FIRMWARE_VOLUME_IMAGE, EFI_SECTION_RAW):
-                    sec.children = build_efi_model( _uefi, sec.Image[sec.HeaderSize:], fwtype )
+                elif sec.Type == EFI_SECTION_FIRMWARE_VOLUME_IMAGE:
+                    children = build_efi_file_tree( _uefi, sec.Image[sec.HeaderSize:], fwtype )
+                    if not children is None:
+						sec.children = children
```
Comment (ErikBjorge) [inline chipsec/hal/spi_uefi.py]: This line seems to have some odd indentation.

File: chipsec/hal/spi_uefi.py [hunk 2]
```diff
@@ -284,18 +284,48 @@ def build_efi_modules_tree( _uefi, fwtype, data, Size, offset, polarity ):
...
+            elif fwbin.Type == EFI_FV_FILETYPE_RAW:
+                if fwbin.Name != NVAR_NVRAM_FS_FILE:
+                    fwbin.children = build_efi_tree( _uefi, fwbin.Image[fhdrsz:], fwtype )
+                    fv.append(fwbin)
+                else:
+                    fwbin.isNVRAM   = True
+                    fwbin.NVRAMType = FWType.EFI_FW_TYPE_NVAR
+                    fv.append(fwbin)
+				#fv.children.append(fwbin)
+		#fv.append(fwbin)
```
Comment (ErikBjorge) [inline chipsec/hal/spi_uefi.py]: Remove commented out code

File: chipsec/hal/spi_uefi.py [hunk 3]
```diff
@@ -310,26 +340,13 @@ def build_efi_tree( _uefi, data, fwtype ):
     while fv_off is not None:
         fv = EFI_FV( fv_off, fv_guid, fv_size, fv_attr, fv_hdrsz, fv_csum, fv_hdroff, fv_img, fv_calccsum )
         fv.calc_hashes()
-        polarity = bit_set( fv.Attributes, EFI_FVB2_ERASE_POLARITY )
+        #polarity = bit_set( fv.Attributes, EFI_FVB2_ERASE_POLARITY )
```
Comment (ErikBjorge) [inline chipsec/hal/spi_uefi.py]: Remove commented out code

File: chipsec/hal/spi_uefi.py [hunk 4]
```diff
@@ -310,26 +340,13 @@ def build_efi_tree( _uefi, data, fwtype ):
...
-                            fwbin.children = build_efi_tree( _uefi, fwbin.Image, fwtype )
-                        else:
-                            fwbin.isNVRAM   = True
-                            fwbin.NVRAMType = FWType.EFI_FW_TYPE_NVAR
-
-                    fv.children.append(fwbin)
-                foff, next_offset, fname, ftype, fattr, fstate, fcsum, fsz, fimg, fhdrsz, fUD, fcalcsum = NextFwFile( fv.Image, fv.Size, next_offset, polarity )
+			fwbin = build_efi_file_tree ( _uefi, fv_img, fwtype)
+			for i in fwbin:
+			    fv.children.append(i)
```
Comment (ErikBjorge) [inline chipsec/hal/spi_uefi.py]: Odd indentation again

File: chipsec/hal/uefi_common.py [hunk 5]
```diff
@@ -623,63 +623,128 @@ def NextFwVolume(buffer, off = 0):
...
+    print "\tHeaderLength:     0x%04X" % HeaderLength
+    print "\tChecksum:         0x%04X" % Checksum
+    print "\tRevision:         0x%02X" % Revision
+    print "\tExtHeaderOffset:  0x%02X" % ExtHeaderOffset
+    print "\tReserved:         0x%02X" % Reserved
+    print "FFS Guid:     %s" % guid_str(FileSystemGuid0, FileSystemGuid1,FileSystemGuid2, FileSystemGuid3)
+    '''
+    while not (numblocks == 0 and lenblock == 0):
+        fof += struct.calcsize(EFI_FV_BLOCK_MAP_ENTRY)
+        if (fof + struct.calcsize(EFI_FV_BLOCK_MAP_ENTRY)) >= len(buffer):
```
Comment (ErikBjorge) [inline chipsec/hal/uefi_common.py]: Would it be better to just calculate the size once and reuse it.  We seem to be calling calcsize in a number of places for the same data structure.

File: chipsec/hal/uefi_common.py [hunk 6]
```diff
@@ -623,63 +623,128 @@ def NextFwVolume(buffer, off = 0):
...
+			return (0,0,0)
+        if numblocks != 0:
+            #print "Num blocks:   0x%08X\n" % (numblocks)
+            #print "block Len:    0x%08X\n" % (lenblock)
+            size = size + (numblocks * lenblock)
+        numblocks,lenblock = struct.unpack(EFI_FV_BLOCK_MAP_ENTRY,buffer[fof:fof+struct.calcsize(EFI_FV_BLOCK_MAP_ENTRY)])
+    if FvLength != size:
+        #print("ERROR: Volume Size not consistant with Block Maps")
```
Comment (ErikBjorge) [inline chipsec/hal/uefi_common.py]: Convert to a DEBUG logger type.

File: chipsec/hal/uefi_common.py [hunk 7]
```diff
@@ -623,63 +623,128 @@ def NextFwVolume(buffer, off = 0):
...
+        return (0,0,0)
+    if size >= 0x40000000 or size == 0:
+        #print("ERROR: Volume is corrupted")
```
Comment (ErikBjorge) [inline chipsec/hal/uefi_common.py]: Convert to a DEBUG log type

File: chipsec/hal/uefi_common.py [hunk 8]
```diff
@@ -623,63 +623,128 @@ def NextFwVolume(buffer, off = 0):
...
+        CalcSum = (hsum | (fsum << 8))
+        res = (cur_offset, next_offset, Name, Type, Attributes, State, IntegrityCheck, fsize, FvImage[fof:fof+fsize], header_size, update_or_deleted, CalcSum)
     if res is None: return (cur_offset, next_offset, None, None, None, None, None, None, None, None, update_or_deleted, None)
     else:           return res
 
 EFI_COMMON_SECTION_HEADER = "<3sB"
 EFI_COMMON_SECTION_HEADER_size = struct.calcsize(EFI_COMMON_SECTION_HEADER)
 
 def NextFwFileSection(sections, ssize, sof, polarity):
+    #print "InNextFileSection",type(sof)
```
Comment (ErikBjorge) [inline chipsec/hal/uefi_common.py]: Remove

**Comment by BrentHoltsclaw** [comment]
@johnloucaides @abazhaniuk any feedback?

**Pull request merged**